### PR TITLE
Upgrade voting to 0.14

### PR DESCRIPTION
**Pull request by orkunkl**
Related to https://github.com/CosmWasm/cosmwasm-examples/issues/121

### CODE REVIEW

File: contracts/voting/.cargo/config [hunk 1]
```diff
@@ -1,5 +1,4 @@
 [alias]
 wasm = "build --release --target wasm32-unknown-unknown"
 unit-test = "test --lib"
-integration-test = "test --test integration"
```
Comment (maurolacy) [inline contracts/voting/.cargo/config:4]: Why are you removing this?

File: contracts/voting/src/msg.rs [hunk 2]
```diff
@@ -35,13 +35,13 @@ pub enum HandleMsg {
...
     Config {},
-    TokenStake { address: HumanAddr },
+    TokenStake { address: String },
     Poll { poll_id: u64 },
 }
 
 #[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema)]
 pub struct PollResponse {
-    pub creator: HumanAddr,
+    pub creator: Addr,
```
Comment (maurolacy) [inline contracts/voting/src/msg.rs:44]: This should be a `String`, as it's part of a query response.

File: contracts/voting/src/contract.rs [hunk 3]
```diff
@@ -17,42 +17,44 @@ const MIN_STAKE_AMOUNT: u128 = 1;
...
     _env: Env,
     info: MessageInfo,
-    msg: InitMsg,
-) -> Result<InitResponse, ContractError> {
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
     let state = State {
         denom: msg.denom,
-        owner: deps.api.canonical_address(&info.sender)?,
+        owner: deps.api.addr_canonicalize(&info.sender.as_str())?,
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:29]: Modify `State` to change from `CanonicalAddr` to`Addr`, and use `info.sender` directly.

File: contracts/voting/src/contract.rs [hunk 4]
```diff
@@ -73,30 +75,30 @@ pub fn stake_voting_tokens(
     deps: DepsMut,
     _env: Env,
     info: MessageInfo,
-) -> Result<HandleResponse, ContractError> {
-    let sender_address_raw = deps.api.canonical_address(&info.sender)?;
+) -> Result<Response, ContractError> {
+    let sender_address_raw = deps.api.addr_canonicalize(&info.sender.as_str())?;
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:79]: Use `info.sender` directly.

File: contracts/voting/src/contract.rs [hunk 5]
```diff
@@ -105,9 +107,9 @@ pub fn withdraw_voting_tokens(
     env: Env,
     info: MessageInfo,
     amount: Option<Uint128>,
-) -> Result<HandleResponse, ContractError> {
-    let sender_address_raw = deps.api.canonical_address(&info.sender)?;
-    let contract_address_raw = deps.api.canonical_address(&env.contract.address)?;
+) -> Result<Response, ContractError> {
+    let sender_address_raw = deps.api.addr_canonicalize(&info.sender.as_str())?;
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:111]: Idem.

File: contracts/voting/src/contract.rs [hunk 6]
```diff
@@ -118,23 +120,22 @@ pub fn withdraw_voting_tokens(
...
             let mut state = config(deps.storage).load()?;
-            let staked_tokens = (state.staked_tokens - withdraw_amount)?;
+            let staked_tokens = state.staked_tokens.checked_sub(withdraw_amount)?;
             state.staked_tokens = staked_tokens;
             config(deps.storage).save(&state)?;
 
             send_tokens(
                 deps.as_ref(),
                 &contract_address_raw,
-                &sender_address_raw,
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:137]: This is wrong. You should remove `contract_address_raw`, not `sender_addres_raw`. Wonder how the tests pass. Maybe it is a good idea to add a test that fails here.

File: contracts/voting/src/contract.rs [hunk 7]
```diff
@@ -105,9 +107,9 @@ pub fn withdraw_voting_tokens(
     env: Env,
     info: MessageInfo,
     amount: Option<Uint128>,
-) -> Result<HandleResponse, ContractError> {
-    let sender_address_raw = deps.api.canonical_address(&info.sender)?;
-    let contract_address_raw = deps.api.canonical_address(&env.contract.address)?;
+) -> Result<Response, ContractError> {
+    let sender_address_raw = deps.api.addr_canonicalize(&info.sender.as_str())?;
+    let contract_address_raw = deps.api.addr_canonicalize(&env.contract.address.as_str())?;
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:112]: This is not needed anymore (but see error below).

File: contracts/voting/src/contract.rs [hunk 8]
```diff
@@ -203,7 +204,7 @@ pub fn create_poll(
     let poll_id = poll_count + 1;
     state.poll_count = poll_id;
 
-    let sender_address_raw = deps.api.canonical_address(&info.sender)?;
+    let sender_address_raw = deps.api.addr_canonicalize(&info.sender.as_str())?;
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:207]: Idem. Use `info.sender` directly.

File: contracts/voting/src/contract.rs [hunk 9]
```diff
@@ -221,11 +222,12 @@ pub fn create_poll(
...
     config(deps.storage).save(&state)?;
 
-    let r = HandleResponse {
+    let r = Response {
+        submessages: vec![],
         messages: vec![],
         attributes: vec![
             attr("action", "create_poll"),
-            attr("creator", deps.api.human_address(&new_poll.creator)?),
+            attr("creator", deps.api.addr_humanize(&new_poll.creator)?),
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:230]: Change `creator` to `Addr`, and use `to_string()` here.

File: contracts/voting/src/contract.rs [hunk 10]
```diff
@@ -244,11 +246,11 @@ pub fn end_poll(
     env: Env,
     info: MessageInfo,
     poll_id: u64,
-) -> Result<HandleResponse, ContractError> {
+) -> Result<Response, ContractError> {
     let key = &poll_id.to_be_bytes();
     let mut a_poll = poll(deps.storage).load(key)?;
 
-    let sender_address_raw = deps.api.canonical_address(&info.sender)?;
+    let sender_address_raw = deps.api.addr_canonicalize(&info.sender.as_str())?;
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:253]: Just use `info.sender`.

File: contracts/voting/src/contract.rs [hunk 11]
```diff
@@ -379,8 +382,8 @@ pub fn cast_vote(
     poll_id: u64,
     vote: String,
     weight: Uint128,
-) -> Result<HandleResponse, ContractError> {
-    let sender_address_raw = deps.api.canonical_address(&info.sender)?;
+) -> Result<Response, ContractError> {
+    let sender_address_raw = deps.api.addr_canonicalize(&info.sender.as_str())?;
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:386]: Idem.

File: contracts/voting/src/contract.rs [hunk 12]
```diff
@@ -431,19 +435,17 @@ pub fn cast_vote(
 
 fn send_tokens(
     deps: Deps,
-    from_address: &CanonicalAddr,
     to_address: &CanonicalAddr,
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:438]: Should be `Addr`. Subsequent changes apply.

If you start by changing `CanonicalAddr` -> `Addr` in `state.rs`, and `HumanAddr` to `String` in `msg.rs` and `query.rs`, than all the changes follow more or less naturally.

**Review by maurolacy — CHANGES_REQUESTED** [review]
Will finish reviewing after requested changes.

File: contracts/voting/.cargo/config (hunk 1 above)
Comment (orkunkl) [reply contracts/voting/.cargo/config:4]: Could not make it work with integration tests, also [vm::MockApi](https://github.com/CosmWasm/cosmwasm/blob/main/packages/vm/src/testing/mock.rs#L65) Is not updated.
Also the integration test are usually same as unit tests in cosmwasm/examples. This makes 2x time to maintain. Actually thinking to remove them?

File: contracts/voting/src/contract.rs (hunk 6 above)
Comment (orkunkl) [reply contracts/voting/src/contract.rs:137]: Actually my local tests were failing, and fixed it now

**Review by maurolacy — APPROVED** [review]
lgtm.

File: contracts/voting/.cargo/config (hunk 1 above)
Comment (maurolacy) [reply contracts/voting/.cargo/config:4]: > Actually thinking to remove them?

OK with me.

File: contracts/voting/src/contract.rs [hunk 13]
```diff
@@ -480,8 +470,8 @@ fn query_poll(deps: Deps, poll_id: u64) -> StdResult<Binary> {
     to_binary(&resp)
 }
 
-fn token_balance(deps: Deps, address: HumanAddr) -> StdResult<Binary> {
-    let key = deps.api.canonical_address(&address).unwrap();
+fn token_balance(deps: Deps, address: Addr) -> StdResult<Binary> {
+    let key = deps.api.addr_canonicalize(&address.as_str())?;
```
Comment (maurolacy) [inline contracts/voting/src/contract.rs:474]: This can be easily removed.

File: contracts/voting/src/tests.rs [hunk 14]
```diff
@@ -142,13 +138,13 @@ mod tests {
...
         let mut deps = mock_dependencies(&[]);
-        mock_init(deps.as_mut());
+        mock_instantiate(deps.as_mut());
         let (env, info) = mock_info_height(TEST_CREATOR, &coins(2, VOTING_TOKEN), 0, 10000);
 
         let quorum = 30;
         let msg = create_poll_msg(quorum, "test".to_string(), None, None);
 
-        let handle_res = handle(deps.as_mut(), env.clone(), info.clone(), msg.clone()).unwrap();
+        let handle_res = execute(deps.as_mut(), env.clone(), info.clone(), msg.clone()).unwrap();
```
Comment (maurolacy) [inline contracts/voting/src/tests.rs:147]: This `handle_res` and subsequent ones can be renamed for clarity.

**Pull request merged**